src/vos/tests/vts_mvcc.c

```diff
@@ -1459,7 +1459,12 @@ uncertainty_check_exec_one(struct io_test_args *arg, int i, int j, bool empty,
 				expected_arc = -DER_EXIST;
 		}
 	}
-	if (is_punch(w) && we > bound)
+	if (is_punch(a) && expected_arc != -DER_NONEXIST && we > ae &&
+	    a->o_wlevel <= w->o_wlevel) {
+		expected_arc = -DER_TX_RESTART;
+	}
```

Comment:
Hmm, it's getting harder and harder to follow, isn't it... Perhaps we should just skip the last case in uncertainty_check_exec when is_punch(a)?

Reply:
OK, finally got it (I hope). we >= ae is always true in this test. How about removing that, flip the negation, recognize that if we <= bound nothing changes, and add a comment, like:

    if (we <= bound) {
        expected_arc = -DER_TX_RESTART;
    } else {
        if ((is_r(a) || is_rw(a))) {
            if (a->o_rtype == R_NE && empty)
                expected_arc = -DER_NONEXIST;
            else if (a->o_rtype == R_E && !empty)
                expected_arc = -DER_EXIST;
        }
        /*
         * If is_punch(a), a is equal to or a parent of
         * w, and the punch part will execute, then
         * a will be rejected due to vos's prevention
         * of "under punches".
         */
        if (is_punch(a) && a->o_wlevel <= w->o_wlevel &&
            expected_arc == 0)
            expected_arc = -DER_TX_RESTART;
    }

Reply:
Yeah, that's a bit simpler.  Done

Reply:
Thanks.